protocol/contracts/governance/ElderElection.sol

```diff
@@ -0,0 +1,143 @@
+ * actioned is handled at the social/policy layer.
+ * 
+ * Any address can vote by endorsing candidates. However only endorsements from voters
+ * who hold temple (aka "token stakeholders") or who are temple team members (aka "impact
+ * stakeholders") have vote value.
+ *
+ * EIP712 structured data signing supports 'gasless voting' via a relay
+ */
+contract ElderElection is EIP712, AccessControl {
+
+    bytes32 public constant CAN_NOMINATE = keccak256("CAN_NOMINATE");
+
+    using Counters for Counters.Counter;
+    using SafeERC20 for IERC20;
+
+    /// @notice The NFT ids (ie discord ids) of the Templars that are candidates for
+    /// the ElderElection
+    mapping(uint256 => bool) public candidates;
+
+    /// @notice The number of active candidates
+    uint256 public numCandidates;
+
+    /// @notice endorsements by a given voter
+    mapping(address => uint256[]) public endorsementsBy;
+
+    /// @notice the NFT contract for templars
+    Templar public templars;
+
+    /// @notice Nonces used in relayed voting requests
+    mapping(address => Counters.Counter) public _nonces;
+
+
+    constructor(
+        Templar _templars
+    ) EIP712("ElderElection", "1") {
+        templars = _templars;
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+
+    /**
+     * @notice Nominate the given discord ID a candidate.
+     */
+    function nominate(uint256 discordId) external onlyRole(CAN_NOMINATE) {
+        templars.checkExists(discordId);
+
+        if (!candidates[discordId]) {
+          candidates[discordId] = true;
+          numCandidates += 1;
+          emit UpdateNomination(discordId, true);
+        }
+    }
+
+    /**
+     * @notice Remove the given discord ID as a candidate
+     */
+    function resign(uint256 discordId) external onlyRole(CAN_NOMINATE) {
+        templars.checkExists(discordId);
+
+        if (candidates[discordId]) {
+          candidates[discordId] = false;
+          numCandidates -= 1;
+          emit UpdateNomination(discordId, false);
+        }
+    }
+
+    function setEndorsements(uint256[] calldata discordIds) external {
+        _setEndorsements(msg.sender, discordIds);
+    }
+
+    function _setEndorsements(address account, uint256[] calldata discordIds) internal {
+        if (discordIds.length > numCandidates) {
+          revert TooManyEndorsements();
+        }
+
+        endorsementsBy[account] = discordIds;
+        emit UpdateEndorsements(account, discordIds);
+    }
+
+    /**
+     * @notice Set the endorsements for an account via relayed, signed request. 
+     */
+    function relayedSetEndorsementsFor(address account, uint256[] calldata discordIds, uint256 deadline, bytes calldata signature) external {
+        if (block.timestamp > deadline) revert DeadlineExpired(deadline);
+
+        bytes32 structHash = keccak256(abi.encode(SET_ENDORSEMENTS_FOR_TYPEHASH, account, discordIds, deadline, _useNonce(account)));
```

Comment:
If the signature has a deadline, is there any value in including a nonce? (I would like to not have to fetch the current nonce before signing on the FE)

Reply:
@butlerji this was from your original implementation. Is the nonce necessary?

Reply:
Defence in depth on my part. 

Timeout stops replay attacks if the tx fails (not perfect, but it's the best we can do)

Nonce stops replay attacks if the Tx is successful.

Mild preference to leave both for the meta reason others will copy/paste this code and use it in other contexts (and also, because it's all public, to stop any mis-guided FUD).

It is mild however, so happy to have just timeout if it makes it materially less effort for ya'll

Reply:
definitely would make it easier to drop the nonces from the signature -- I don't see an effective replay attack here

Reply:
Would this be a possible replay attack:

   - User creates a signed TX1 to set endorsements via the relay
   - User changes endorsements be creating a signed TX2 and sending to the relay
   - Hacker undoes TX2 by replaying TX1

Reply:
that's true, if they undo their endorsement before the deadline then the first tx could be replayed